src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/ExternalProcessRunner.java

```diff
@@ -0,0 +1,117 @@
+ * THE SOFTWARE.
+ */
+package org.niis.xroad.restapi.service;
+
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.compress.utils.IOUtils;
+import org.springframework.stereotype.Component;
+import org.springframework.util.StringUtils;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+@Slf4j
+@Component
+public class ExternalProcessRunner {
+    private static final long TIMEOUT = 60000;
+
+    /**
+     * Executes the given command with given arguments.
+     * @param command the command to execute
+     * @param args arguments to be appended to the command. Make sure to pass your arguments in the correct order
+     * (e.g. if your options have values enter them as separate consecutive args).
+     * @return the output of the executed process as a List of Strings
+     * @throws ProcessNotExecutableException in the case of IOException or if the process is interrupted
+     * @throws ProcessFailedException if the process' exit code is not 0
+     */
+    public List<String> execute(String command, String... args) throws ProcessNotExecutableException,
+            ProcessFailedException, InterruptedException {
+        if (StringUtils.isEmpty(command)) {
+            throw new IllegalArgumentException("command cannot be null");
+        }
+
+        List<String> commandWithArgs = new ArrayList<>();
+        commandWithArgs.add(command);
+        if (args != null && args.length > 0) {
+            commandWithArgs.addAll(Arrays.asList(args));
+        }
+        Process process;
+        ProcessBuilder pb = new ProcessBuilder(commandWithArgs);
+        // redirect process errors into process's input stream
+        pb.redirectErrorStream(true);
+        try {
+            process = pb.start();
+        } catch (IOException e) {
+            throw new ProcessNotExecutableException(e);
+        }
+
+        // gather output into a list of string for returning
+        List<String> processOutput = new ArrayList<>();
+        try (BufferedReader br = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
+            br.lines().forEach(processOutput::add);
+        } catch (IOException e) {
+            process.destroy();
+            IOUtils.closeQuietly(process.getErrorStream());
+            IOUtils.closeQuietly(process.getOutputStream());
+            throw new ProcessNotExecutableException(e);
+        }
+
+        int exitCode;
+
+        try {
+            boolean hasExited = process.waitFor(TIMEOUT, TimeUnit.MILLISECONDS);
+            // exit value cannot be asked if the process is still running after timeout - instead throw and destroy
+            if (!hasExited) {
+                throw new ProcessFailedException("Process timed out");
+            }
+            exitCode = process.exitValue();
+        } catch (InterruptedException e) {
```

Comment:
the comment "we dont want..." seems outdated

Reply:
Comment updated